src/utils/str.c

```diff
@@ -56,110 +56,198 @@ void str_printf(str *dst, const char *format, ...) {
     // vsnprintf may return -1 for errors, catch that here.
     if(size < 0) {
         PERROR("Call to vsnprintf returned -1");
-        exit(1);
+        abort();
     }
 
     // Make sure there is enough room for our vsnprintf call plus ending NULL,
-    // then render the output to our new buffer.
-    dst->data = omf_realloc(dst->data, dst->len + size + 1);
-    vsnprintf(dst->data + dst->len, size + 1, format, args2);
+    // then render the output to our new buffer. Note that "size" includes the
+    // terminating null, so we pass "size-1" for allocator which wants the size
+    // of the string without any trailing nulls characters. Since vsnprintf DOES
+    // want to also know about the trailing null, we pass it the original "size"
+    // as is. As a result, we have a proper C string in dst->data, no need to
+    // set trailing null by ourselves anymore.
+    STR_ALLOC(dst, size - 1);
+    vsnprintf(dst->data, size, format, args2);
     va_end(args2);
-
-    dst->len += size;
-    dst->data[dst->len] = 0;
 }
 
-void str_slice(str *dst, const str *src, size_t start, size_t end) {
+void str_from_slice(str *dst, const str *src, size_t start, size_t end) {
+    assert(dst != src);
+    assert(start >= 0);
+    assert(end >= 0);
```

Comment:
start and end are size_t, so these asserts cannot be false

Reply:
Righto, fixed.